mono/mini/mini-posix.c

```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
+	 * simply return. If not, we calculate the difference and do another sleep.
+	 *
+	 * This should reduce the amount of drift that happens because we account
+	 * for the time spent executing the signal handler, which nanosleep () is
+	 * not guaranteed to do for the rem argument.
+	 *
+	 * The downside to this approach is that it is slightly expensive: We have
+	 * to make an extra system call to retrieve the current time whenever we're
+	 * going to restart a nanosleep () call. This is unlikely to be a problem
+	 * in practice since the sampling thread won't be receiving many signals in
+	 * the first place (it's a tools thread, so no STW), and because typical
+	 * sleep periods for the thread are many orders of magnitude bigger than
+	 * the time it takes to actually perform that system call (just a few
+	 * nanoseconds).
+	 */
+	do {
+		gint64 diff = (gint64) ns_abs - (gint64) clock_get_time_ns (mode);
+
+		if (diff <= 0)
+			break;
+
+		req.tv_sec = diff / 1000000000;
+		req.tv_nsec = diff % 1000000000;
+
+		if ((ret = nanosleep (&req, NULL)) == -1 && errno != EINTR)
+			g_error ("%s: nanosleep () returned -1, errno = %d", __func__, errno);
+	} while (ret == -1);
 #endif
+
+	MONO_FINISH_BLOCKING;
 }
 
-#ifdef ITIMER_PROF
-static int
-get_itimer_mode (void)
+static int profiler_signal;
+static MonoNativeThreadId sampling_thread;
+static volatile gint32 sampling_thread_running;
+
+static mono_native_thread_return_t
+sampling_thread_func (void *data)
 {
-	switch (mono_profiler_get_sampling_mode ()) {
-	case MONO_PROFILER_STAT_MODE_PROCESS: return ITIMER_PROF;
-	case MONO_PROFILER_STAT_MODE_REAL: return ITIMER_REAL;
+	mono_threads_attach_tools_thread ();
+
+	MonoProfileSamplingMode mode = mono_profiler_get_sampling_mode ();
+	gint64 rate = 1000000000 / mono_profiler_get_sampling_rate ();
+
+	int old_policy;
+	struct sched_param old_sched;
+	pthread_getschedparam (pthread_self (), &old_policy, &old_sched);
+
+	/*
+	 * Attempt to switch the thread to real time scheduling. This will not
+	 * necessarily work on all OSs; for example, most Linux systems will give
+	 * us EPERM here unless configured to allow this.
+	 *
+	 * TODO: This does not work on Mac (and maybe some other OSs). On Mac, we
+	 * have to use the Mach thread policy routines to switch to real-time
+	 * scheduling. This is quite tricky as we need to specify how often we'll
+	 * be doing work (easy), the normal processing time needed (also easy),
+	 * and the maximum amount of processing time needed (hard). This is
+	 * further complicated by the fact that if we misbehave and take too long
+	 * to do our work, the kernel may knock us back down to the normal thread
+	 * scheduling policy without telling us.
+	 */
+	struct sched_param sched = { .sched_priority = sched_get_priority_max (SCHED_FIFO) };
+	pthread_setschedparam (pthread_self (), SCHED_FIFO, &sched);
+
+	while (InterlockedRead (&sampling_thread_running)) {
+		guint64 sleep = clock_get_time_ns (mode) + rate;
```

Comment:
This is explicitly introducing signaling drift back.

We should compute the time once and simply add rate on each loop.